modules/file-transfer/src/main/java/org/apache/ignite/internal/network/file/FileTransferServiceImpl.java

```diff
@@ -244,69 +254,84 @@ private void processFileTransferInitMessage(FileTransferInitMessage message, Str
 
         Identifier identifier = message.identifier();
 
-        CompletableFuture<Path> directoryFuture = supplyAsync(() -> createTransferDirectory(transferId), executorService);
+        CompletableFuture<Path> directoryFuture = supplyAsync(() -> createTransferDirectory(transferId), executorService)
+                .whenComplete((directory, e) -> {
+                    if (e != null) {
+                        LOG.error("Failed to create transfer directory [transferId={}, identifier={}]", e, transferId, identifier);
+                    }
+                });
 
         CompletableFuture<List<Path>> uploadedFiles = directoryFuture.thenCompose(directory -> fileReceiver.registerTransfer(
                 senderConsistentId,
                 transferId,
                 message.headers(),
                 directory
-        )).whenComplete((v, e) -> {
-            if (e != null) {
-                LOG.error("Failed to register file transfer. Transfer ID: {}. Metadata: {}", e, transferId, identifier);
-            }
-        });
-
-        FileTransferInitResponse response = messageFactory.fileTransferInitResponse().build();
-
-        messagingService.respond(senderConsistentId, FILE_TRANSFER_CHANNEL, response, correlationId)
+        ));
+
+        // Check that directory was created successfully and send response to the sender.
+        directoryFuture.handle((directory, e) -> {
```

Comment:
I know you're tired of me at this point, but does it really have to be one huge statement in the code?
Why not split it into several statements, by extracting temporary results into variables?
Why not extract some parts into methods?
Are you comfortable yourself with the code of such complexity?